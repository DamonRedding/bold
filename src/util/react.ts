import { MutableRefObject, Ref } from 'react'

export function composeRefs<T>(...refs: Array<Ref<T>>) {
    return (value: T) => refs.forEach(ref => setRef(ref, value))
}

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}

/**
 * Returns the next sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getNextSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.nextElementSibling) {
        current = current.nextElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}

/**
 * Returns the previous sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getPreviousSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.previousElementSibling) {
        current = current.previousElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}
