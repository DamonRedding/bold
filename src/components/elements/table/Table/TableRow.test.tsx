import { mount, render } from 'enzyme'
import * as React from 'react'

import { withTheme } from '../../../../test'

import { Table, TableBody, TableCell, TableHead, TableHeader, TableRow } from './index'

describe('TableRow', () => {
    it('should have the pointer class if onClick is specified', () => {
        expect(render(withTheme(
            <Table>
                <TableBody>
                    <TableRow><TableCell>Test</TableCell></TableRow>
                </TableBody>
            </Table>
        ))).toMatchSnapshot()

        const clickHandler = jest.fn()
        const wrapper = mount(withTheme(
            <Table>
                <TableBody>
                    <TableRow onClick={clickHandler}><TableCell>Test</TableCell></TableRow>
                </TableBody>
            </Table>
        ))
        expect(wrapper.render()).toMatchSnapshot()

        wrapper.find('tr').simulate('click')
        expect(clickHandler).toHaveBeenCalled()
    })
    it('should accept the style prop', () => {
        expect(render(withTheme(
            <Table>
                <TableBody>
                    <TableRow style={{ color: 'red' }}><TableCell>Test</TableCell></TableRow>
                </TableBody>
            </Table>
        ))).toMatchSnapshot()
    })
})
