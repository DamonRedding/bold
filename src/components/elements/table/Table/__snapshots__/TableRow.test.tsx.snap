// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow should accept the style prop 1`] = `
.emotion-3 {
  width: 100%;
  border-collapse: collapse;
  font-size: 0.75rem;
  border: 1px solid #e6e6e6;
}

.emotion-0 {
  text-align: left;
  padding: 0 1rem;
  height: 2.5rem;
  border-bottom: 1px solid #e6e6e6;
}

.emotion-1 {
  color: red;
}

<table
  class="emotion-3"
>
  <tbody
    class="emotion-2"
  >
    <tr
      class="emotion-1"
    >
      <td
        class="emotion-0"
      >
        Test
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TableRow should have the pointer class if onClick is specified 1`] = `
.emotion-3 {
  width: 100%;
  border-collapse: collapse;
  font-size: 0.75rem;
  border: 1px solid #e6e6e6;
}

.emotion-0 {
  text-align: left;
  padding: 0 1rem;
  height: 2.5rem;
  border-bottom: 1px solid #e6e6e6;
}

<table
  class="emotion-3"
>
  <tbody
    class="emotion-1"
  >
    <tr
      class="emotion-1"
    >
      <td
        class="emotion-0"
      >
        Test
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`TableRow should have the pointer class if onClick is specified 2`] = `
.emotion-3 {
  width: 100%;
  border-collapse: collapse;
  font-size: 0.75rem;
  border: 1px solid #e6e6e6;
}

.emotion-0 {
  text-align: left;
  padding: 0 1rem;
  height: 2.5rem;
  border-bottom: 1px solid #e6e6e6;
}

.emotion-1 {
  cursor: pointer;
}

<table
  class="emotion-3"
>
  <tbody
    class="emotion-2"
  >
    <tr
      class="emotion-1"
    >
      <td
        class="emotion-0"
      >
        Test
      </td>
    </tr>
  </tbody>
</table>
`;
